packages/thrift-server-core/src/main/Int64.ts

```diff
@@ -150,91 +322,57 @@ export class Int64 {
     public setValue(str: string): void
     public setValue(hi: number | string, lo?: number): void
     public setValue(hi: any, lo?: any): void {
-        let negate: boolean = false
         if (lo === undefined) {
             if (typeof hi === 'number') {
-                // Simplify bitfield retrieval by using abs() value.  We restore sign
-                // later
-                negate = hi < 0
-                hi = Math.abs(hi)
-                lo = hi % VAL32
-                hi = hi / VAL32
-                if (hi > VAL32) {
-                    throw new RangeError(hi + ' is outside Int64 range')
-                } else {
-                    hi = hi | 0
-                }
-            } else if (typeof hi === 'string') {
-                hi = (hi + '').replace(/^0x/, '')
-                lo = hi.substr(-8)
-                hi = hi.length > 8 ? hi.substr(0, hi.length - 8) : ''
-                hi = parseInt(hi, 16)
-                lo = parseInt(lo, 16)
-            } else {
-                throw new Error(hi + ' must be a Number or String')
+                this.setNumber(hi)
+                return
             }
+            if (typeof hi === 'string') {
+                this.setHexString(hi)
+                return
+            }
+            throw new Error(hi + ' must be a Number or String')
         }
 
-        // Technically we should throw if hi or lo is outside int32 range here, but
-        // it's not worth the effort. Anything past the 32'nd bit is ignored.
-
-        // Copy bytes to buffer
-        const b: Buffer = this.buffer
-        const o: number = this.offset
-        for (let i = 7; i >= 0; i--) {
-            b[o + i] = lo & 0xff
-            lo = i === 4 ? hi : lo >>> 8
+        if (typeof hi !== 'number' || typeof lo !== 'number') {
+            throw new Error(`${hi} and ${lo} must be Numbers`)
         }
 
-        // Restore sign of passed argument
-        if (negate) {
-            this._2scomp()
-        }
+        this.setHiLo({ hi, lo })
     }
 
+    /** @inheritDoc */
     public toDecimalString(): string {
-        const i64: Int64 = this
-        let b = i64.buffer
-        const o = i64.offset
+        // Get the number from the buffer.
+        const i64 = this.read()
 
-        if ((!b[o] && !(b[o + 1] & 0xe0)) || (!~b[o] && !~(b[o + 1] & 0xe0))) {
```

Comment:
All bitwise operations on `number` are 32-bit. The bitwise `~` operator is not working as intended for negative numbers.

The real check should be `b[o] === 0xff` but the `!~0xff` results in `!0xffffff00` which is `false`.

Reply:
The test value `-123` was not considered "magnitude is small enough" which ends up assuming the number is more than 53-bits so it ends up zero padding the value quite a bit.